File: schemas/booking-scheduling/mongodb/locations.js
// locations: Physical and virtual venues where services are offered.
// See README.md for full design rationale.

const mongoose = require("mongoose");

const locationsSchema = new mongoose.Schema(
  {
    name: { type: String, required: true },
    slug: { type: String, unique: true, required: true },
    type: { type: String, enum: ["physical", "virtual"], required: true, default: "physical" },
    description: { type: String, default: null },
    address_line1: { type: String, default: null },
    address_line2: { type: String, default: null },
    city: { type: String, default: null },
    state: { type: String, default: null },
    postal_code: { type: String, default: null },
    country: { type: String, default: null },
    virtual_url: { type: String, default: null },
    timezone: { type: String, required: true },
    phone: { type: String, default: null },
    email: { type: String, default: null },
    is_active: { type: Boolean, required: true, default: true },
    position: { type: Number, required: true, default: 0 },
  },
  { timestamps: { createdAt: "created_at", updatedAt: "updated_at" } }
);

locationsSchema.index({ type: 1 });
locationsSchema.index({ is_active: 1, position: 1 });

module.exports = mongoose.model("Location", locationsSchema);

File: schemas/booking-scheduling/spacetimedb/locations.rs
// locations: Physical or virtual venues where services are offered.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum LocationType {
    Physical, // type: String
    Virtual,
}

#[spacetimedb::table(name = locations, public)]
pub struct Location {
    #[primary_key]
    pub id: String, // UUID
    pub name: String,
    #[unique]
    pub slug: String,
    #[index(btree)]
    pub location_type: LocationType,
    pub description: Option<String>,
    pub address_line1: Option<String>,
    pub address_line2: Option<String>,
    pub city: Option<String>,
    pub state: Option<String>,
    pub postal_code: Option<String>,
    pub country: Option<String>,
    pub virtual_url: Option<String>,
    pub timezone: String,
    pub phone: Option<String>,
    pub email: Option<String>,
    #[index(btree)]
    pub is_active: bool,
    pub position: i32,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
    // Composite index: (is_active, position)
}
